pytensor/link/jax/linker.py

```diff
@@ -12,8 +12,28 @@ class JAXLinker(JITLinker):
 
     def fgraph_convert(self, fgraph, input_storage, storage_map, **kwargs):
         from pytensor.link.jax.dispatch import jax_funcify
+        from pytensor.sparse.type import SparseTensorType
         from pytensor.tensor.random.type import RandomType
 
+        if any(
+            isinstance(inp.type, RandomType) and not isinstance(inp, SharedVariable)
+            for inp in fgraph.inputs
+        ):
+            warnings.warn(
```

Comment:
Similar comment to the one in the tests. Should a user using pytensor to compile to jax really be expected to provide jax-valued inputs? If so, why not just write in JAX directly (instead of passing through pytensor), and if not why issue this warning?

Reply:
I commented above, the point is PyTensor is an abstraction and then tries to lower to specialized backends. It's not always possible to lower from a generic format (in our case numpy-scipy) into specialized ones (like JAX) because they may have types that are fundamentally different. Most inputs tend to be numpy-like things that our backends all support natively so these concerns don't show up. But once we move to stuff like Sparse and RNG variables things break down.

In the spirit of allowing one graph - multiple backends we have to at some point coerce types into the specific ones required by the backends. The other option would be for users to create the specialized graphs with specialized types (which then couldn't be compiled to other types). In that case you get closer to your point of "why not just write in JAX" that this approach actually tries to avoid.

For libraries like PyMC it's quite important that we don't force users to specify the backend in advance. Specially because the standard workflow actually uses different backends for different tasks (e.g. random sampling and mcmc sampling)

Reply:
We could not issue a warning (it's one of the things I am asking feedback about). However, it's clear that when we come around to address shared variables that will have to be a warning, because those will inevitably have to de copied over since they can't override the original variables (we have such a warning for Shared RNGs, which you have probably seen, and is conceptually a bit different than the one being introduced here)